### PR TITLE
Adding Additional Test Cases for JSONDecoder in core/cautils/getter

### PR DESCRIPTION
**Pull request by cbrom**
## PR Type:
Tests

___
## PR Description:
This PR introduces additional test cases for the JSONDecoder function in the core/cautils/getter package. The new tests cover the following scenarios:
- An empty string input to the JSONDecoder, which should return an EOF error.
- An invalid JSON string input to the JSONDecoder, which should return an error.

___
## PR Main Files Walkthrough:
<details> <summary>files:</summary>

- `core/cautils/getter/json_test.go`: Two new test cases have been added:
- The first test checks for an EOF error when an empty string is passed to the JSONDecoder.
- The second test checks for an error when an invalid JSON string is passed to the JSONDecoder.
</details>

### CODE REVIEW

**Comment by matthyx** [comment]
@cbrom is it ready to review?

**Comment by cbrom** [comment]
@matthyx yes.

File: core/cautils/getter/json_test.go [hunk 1]
```diff
@@ -29,4 +29,19 @@ func TestJSONDecoder(t *testing.T) {
...
+	t.Run("empty string input should return EOF error", func(t *testing.T) {
+		const input = ``
+		d := JSONDecoder(input)
+		var receiver string
+		require.Error(t, d.Decode(&receiver))
+		require.Equal(t, "", receiver)
+	})
+
+	t.Run("return error if input string is not a valid json", func(t *testing.T) {
+		const input = `{"name": "John", "age": 30,}`
```
Comment (matthyx) [inline core/cautils/getter/json_test.go:42]: maybe you can put `not a valid json` as the string to clearly show it is not valid

**Comment by matthyx** [comment]
> * **suggestion:** Consider adding a test case for a valid JSON string to ensure that the function also works as expected in normal scenarios. [medium]

nice one

**Comment by matthyx** [comment]
@cbrom can you check the 2 suggestions above?

**Comment by cbrom** [comment]
@matthyx while working on this one, I could not find any usage for this line
https://github.com/kubescape/kubescape/blob/master/core/cautils/getter/json.go#L23
Tests pass whether this line exists or not. What am I missing.

**Comment by cbrom** [comment]
> @matthyx while working on this one, I could not find any usage for this line https://github.com/kubescape/kubescape/blob/master/core/cautils/getter/json.go#L23 Tests pass whether this line exists or not. What am I missing.

oops, nvm. I guess that works with actual JSON input.

**Comment by matthyx** [comment]
yes, actually we should probably refactor https://github.com/kubescape/kubescape/blob/c9ccef90f33c9fc402115f392c55011493de8dfd/core/cautils/versioncheck.go#L141

**Comment by cbrom** [comment]
@matthyx do you have any recommendation/style on how the tesstcases are structured here?

**Comment by cbrom** [comment]
> yes, actually we should probably refactor
> 
> https://github.com/kubescape/kubescape/blob/c9ccef90f33c9fc402115f392c55011493de8dfd/core/cautils/versioncheck.go#L141

`client := getter.GetKSCloudAPIConnector()` any suggestion on how I use this client?

**Comment by matthyx** [comment]
> GetKSCloudAPIConnector

Indeed I think https://github.com/kubescape/kubescape/blob/c4f0e6e46bf03a2c0e2782baed485d603a0d019a/core/cautils/getter/kscloudapi.go is the right place to implement the new logic, however you need to implement this functionnality.

**Comment by cbrom** [comment]
> > GetKSCloudAPIConnector
> 
> Indeed I think https://github.com/kubescape/kubescape/blob/c4f0e6e46bf03a2c0e2782baed485d603a0d019a/core/cautils/getter/kscloudapi.go is the right place to implement the new logic, however you need to implement this functionnality.

@matthyx 
I couldn't find the relevant functions to get the latest version in KSCloudAPI. Even the get and post methods are not exported. Let me know if you have any suggestion please.

**Comment by matthyx** [comment]
> @matthyx I couldn't find the relevant functions to get the latest version in KSCloudAPI. Even the get and post methods are not exported. Let me know if you have any suggestion please.

it doesn't exist... but I was suggesting if you want to implement it that's effectively the place to do it

**Comment by cbrom** [comment]
> > @matthyx I couldn't find the relevant functions to get the latest version in KSCloudAPI. Even the get and post methods are not exported. Let me know if you have any suggestion please.
> 
> it doesn't exist... but I was suggesting if you want to implement it that's effectively the place to do it

@matthyx I have made the necessary changes, would you check if I am on the right path before I go ahead with the tests please?

**Comment by matthyx** [comment]
> > > @matthyx I couldn't find the relevant functions to get the latest version in KSCloudAPI. Even the get and post methods are not exported. Let me know if you have any suggestion please.
> > 
> > 
> > it doesn't exist... but I was suggesting if you want to implement it that's effectively the place to do it
> 
> @matthyx I have made the necessary changes. Would you check if I am on the right path before I proceed with the tests please?

OMG does it have to be that complicated? we just need to do a POST, not write a new mux for http

**Comment by cbrom** [comment]
> > > > @matthyx I couldn't find the relevant functions to get the latest version in KSCloudAPI. Even the get and post methods are not exported. Let me know if you have any suggestion please.
> > > 
> > > 
> > > it doesn't exist... but I was suggesting if you want to implement it that's effectively the place to do it
> > 
> > 
> > @matthyx I have made the necessary changes. Would you check if I am on the right path before I proceed with the tests please?
> 
> OMG does it have to be that complicated? we just need to do a POST, not write a new mux for http

Ah, I guess I did not get your intention and with the comment 
```
// HttpPost provides a low-level capability to send a HTTP POST request and serialize the response as a string.
//
// Deprecated: use methods of the KSCloudAPI client instead.
```

Let me revert my changes then.

**Comment by cbrom** [comment]
@matthyx sorry for the confusion, I have simplified the post request function.

File: core/cautils/getter/kscloudapi.go [hunk 2]
```diff
@@ -45,3 +50,25 @@ func GetKSCloudAPIConnector() *v1.KSCloudAPI {
 
 	return &client
 }
+
+// APIPostRequest provides a low-level utility that sends a POST request to a given url
+func APIPostRequest(fullURL string, body []byte, headers map[string]string) (io.ReadCloser, int64, error) {
```
Comment (matthyx) [inline core/cautils/getter/kscloudapi.go:55]: do you mind adding a test for that? ideally you would start an http listener for the test and verify the correct post and send back an answer

**Review by matthyx — CHANGES_REQUESTED** [review]

File: core/cautils/getter/kscloudapi.go (hunk 2 above)
Comment (cbrom) [reply core/cautils/getter/kscloudapi.go:55]: Yeah, I just wanted to know if I was on the right path first. I'll go ahead and do that.

**Comment by cbrom** [comment]
@matthyx Ready for review.

**Comment by matthyx** [comment]
> @matthyx Ready for review.

can you squash?

**Comment by cbrom** [comment]
> > @matthyx Ready for review.
> 
> can you squash?

done.

**Comment by matthyx** [comment]
> > > @matthyx Ready for review.
> > 
> > 
> > can you squash?
> 
> done.

```
# github.com/kubescape/kubescape/v3/core/cautils/getter
core/cautils/getter/getpoliciesutils.go:21:22: undefined: json
core/cautils/getter/loadpolicy.go:72:11: undefined: json
core/cautils/getter/loadpolicy.go:82:11: undefined: json
core/cautils/getter/loadpolicy.go:110:12: undefined: json
core/cautils/getter/loadpolicy.go:134:12: undefined: json
core/cautils/getter/loadpolicy.go:163:13: undefined: json
core/cautils/getter/loadpolicy.go:189:11: undefined: json
core/cautils/getter/loadpolicy.go:213:8: undefined: json
core/cautils/getter/loadpolicy.go:237:11: undefined: json
core/cautils/getter/loadpolicy.go:258:11: undefined: json
core/cautils/getter/loadpolicy.go:237:11: too many errors
FAIL	github.com/kubescape/kubescape/v3 [build failed]
FAIL	github.com/kubescape/kubescape/v3/cmd/scan [build failed]
FAIL	github.com/kubescape/kubescape/v3/cmd/shared [build failed]
FAIL	github.com/kubescape/kubescape/v3/cmd/version [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/cautils [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/cautils/getter [build failed]
?   	github.com/kubescape/kubescape/v3/core/meta/datastructures/v1	[no test files]
FAIL	github.com/kubescape/kubescape/v3/core/core [build failed]
?   	github.com/kubescape/kubescape/v3/core/metrics	[no test files]
?   	github.com/kubescape/kubescape/v3/core/mocks	[no test files]
ok  	github.com/kubescape/kubescape/v3/core/pkg/containerscan	0.009s
?   	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/gotree	[no test files]
ok  	github.com/kubescape/kubescape/v3/core/pkg/fixhandler	0.144s
FAIL	github.com/kubescape/kubescape/v3/core/pkg/hostsensorutils [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/pkg/opaprocessor [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/pkg/policyhandler [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/pkg/resourcehandler [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/pkg/resourcesprioritization [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling [build failed]
?   	github.com/kubescape/kubescape/v3/internal/testutils	[no test files]
ok  	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/locationresolver	0.155s
FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/printer/v2 [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/printer/v2/prettyprinter [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/printer/v2/prettyprinter/tableprinter/configurationprinter [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/printer/v2/prettyprinter/tableprinter/imageprinter [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/printer/v2/prettyprinter/tableprinter/utils [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/reporter/v2 [build failed]
FAIL	github.com/kubescape/kubescape/v3/core/pkg/score [build failed]
ok  	github.com/kubescape/kubescape/v3/pkg/imagescan	0.022s [no tests to run]
FAIL
```

**Comment by cbrom** [comment]
> > > > @matthyx Ready for review.
> > > 
> > > 
> > > can you squash?
> > 
> > 
> > done.
> 
> ```
> # github.com/kubescape/kubescape/v3/core/cautils/getter
> core/cautils/getter/getpoliciesutils.go:21:22: undefined: json
> core/cautils/getter/loadpolicy.go:72:11: undefined: json
> core/cautils/getter/loadpolicy.go:82:11: undefined: json
> core/cautils/getter/loadpolicy.go:110:12: undefined: json
> core/cautils/getter/loadpolicy.go:134:12: undefined: json
> core/cautils/getter/loadpolicy.go:163:13: undefined: json
> core/cautils/getter/loadpolicy.go:189:11: undefined: json
> core/cautils/getter/loadpolicy.go:213:8: undefined: json
> core/cautils/getter/loadpolicy.go:237:11: undefined: json
> core/cautils/getter/loadpolicy.go:258:11: undefined: json
> core/cautils/getter/loadpolicy.go:237:11: too many errors
> FAIL	github.com/kubescape/kubescape/v3 [build failed]
> FAIL	github.com/kubescape/kubescape/v3/cmd/scan [build failed]
> FAIL	github.com/kubescape/kubescape/v3/cmd/shared [build failed]
> FAIL	github.com/kubescape/kubescape/v3/cmd/version [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/cautils [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/cautils/getter [build failed]
> ?   	github.com/kubescape/kubescape/v3/core/meta/datastructures/v1	[no test files]
> FAIL	github.com/kubescape/kubescape/v3/core/core [build failed]
> ?   	github.com/kubescape/kubescape/v3/core/metrics	[no test files]
> ?   	github.com/kubescape/kubescape/v3/core/mocks	[no test files]
> ok  	github.com/kubescape/kubescape/v3/core/pkg/containerscan	0.009s
> ?   	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/gotree	[no test files]
> ok  	github.com/kubescape/kubescape/v3/core/pkg/fixhandler	0.144s
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/hostsensorutils [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/opaprocessor [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/policyhandler [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/resourcehandler [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/resourcesprioritization [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling [build failed]
> ?   	github.com/kubescape/kubescape/v3/internal/testutils	[no test files]
> ok  	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/locationresolver	0.155s
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/printer/v2 [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/printer/v2/prettyprinter [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/printer/v2/prettyprinter/tableprinter/configurationprinter [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/printer/v2/prettyprinter/tableprinter/imageprinter [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/printer/v2/prettyprinter/tableprinter/utils [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/resultshandling/reporter/v2 [build failed]
> FAIL	github.com/kubescape/kubescape/v3/core/pkg/score [build failed]
> ok  	github.com/kubescape/kubescape/v3/pkg/imagescan	0.022s [no tests to run]
> FAIL
> ```

Fixed. Sorry about that.

**Review by matthyx — APPROVED** [review]

**Pull request merged**